File: Android/shiba/src/main/java/moe/tlaster/shiba/scripting/IScriptRuntime.kt
package moe.tlaster.shiba.scripting

import moe.tlaster.shiba.scripting.conversion.ITypeConversion

interface IScriptRuntime {
    fun hasFunction(name: String): Boolean
    fun hasFunction(instance: Any?, name: String): Boolean
    fun callFunction(name: String, vararg parameters: Any?): Any?
    fun callFunction(instance: Any?, name: String, vararg parameters: Any?): Any?
    fun execute(script: String): Any?
    fun addTypeConversion(conversion: ITypeConversion)
    fun getProperty(instance: Any?, name: String): Any?
    fun isArray(instance: Any?): Boolean
    fun isObject(instance: Any?): Boolean
    fun toArray(instance: Any?): List<Any>
    fun injectFunction(instance: Any?, name: String, block: () -> Unit)
    fun <T: Any> injectFunction(instance: Any?, name: String, block: (T) -> Unit)
}

File: UWP/Shiba/Scripting/IScriptRuntime.cs
using System.Collections;

namespace Shiba.Scripting
{
    public interface IScriptRuntime
    {
        void AddObject(string name, object value);
        bool HasFunction(string name);
        bool HasFunction(object instance, string name);
        object CallFunctionWithCustomThis(object thiz, string functionName, params object[] parameters);
        object CallFunction(string functionName, params object[] parameters);
        object CallFunction(object instance, string functionName, params object[] parameters);
        object GetProperty(object instance, string propertyName);
        object GetAtIndex(object instance, int index);
        void AddRef(object instance);
        void ReleaseObj(object instance);
        bool IsArray(object instance);
        IEnumerable ToArray(object instance);
        object Execute(string script);
    }
}